### PR TITLE
Record: make field ordering deterministic, add test

### PR DESCRIPTION
**Pull request by matthiasdiener**
This is a proof-of-concept and is ready for a first look @inducer. I'm not sure if there is a better way to do this.

_Please squash_

### CODE REVIEW

**Comment by inducer** [comment]
- This seems OK, although it is a compatibility break, which I'm not loving.
- We should just deprecate the thing for removal in 2025. I.e. have it spew warnings, and remove it from the docs. (in favor of dataclasses, which do the same thing, but more competently)

**Comment by matthiasdiener** [comment]
> * This seems OK, although it is a compatibility break, which I'm not loving.

Is there a use case that breaks with this change? My goal was to remain compatible with the previous set-based implementation (which is the reason for the conversion from set to dict in `__init__`).

> * We should just deprecate the thing for removal in 2025. I.e. have it spew warnings, and remove it from the docs. (in favor of dataclasses, which do the same thing, but more competently)

What do you think of [c32e6d0](https://github.com/inducer/pytools/pull/199/commits/c32e6d01def6383a126059a9932f0c320783c7a1)?

**Comment by matthiasdiener** [comment]
Setting this as ready for review. mirgecom seems to run fine with it.

File: pytools/__init__.py [hunk 1]
```diff
@@ -468,8 +476,6 @@ def __getattr__(self, name):
 
 
 class Record(RecordWithoutPickling):
-    __slots__: ClassVar[List[str]] = []
```
Comment (inducer) [inline pytools/__init__.py:471]: Is there a specific reason this was dropped?

File: pytools/__init__.py [hunk 2]
```diff
@@ -411,9 +411,14 @@ class RecordWithoutPickling:
...
 
     __slots__: ClassVar[List[str]] = []
-    fields: ClassVar[Set[str]]
+
+    # A dict, not a set, to maintain a deterministic iteration order
+    fields: ClassVar[Dict[str, None]]
 
     def __init__(self, valuedict=None, exclude=None, **kwargs):
+        from warnings import warn
+        warn("pytools.Record is deprecated. Use dataclasses instead.")
```
Comment (inducer) [inline pytools/__init__.py:420]: Also remove from docs.

Comment (inducer) [inline pytools/__init__.py:420]: Note that this is `RecordWithoutPickling`.

Comment (matthiasdiener) [reply pytools/__init__.py:420]: Is it documented somewhere? I could not find it.

Comment (inducer) [reply pytools/__init__.py:420]: Looks like we already got rid of it. Good! Nevermind then. :slightly_smiling_face:

File: pytools/__init__.py (hunk 1 above)
Comment (matthiasdiener) [reply pytools/__init__.py]: I had overlooked that this must be defined in every subclass to be effective 🤦 . Restored in [0a0fcef](https://github.com/inducer/pytools/pull/199/commits/0a0fcef799d5fc19e1e27541b298d524943197a1)

File: pytools/__init__.py (hunk 2 above)
Comment (matthiasdiener) [reply pytools/__init__.py:420]: Right, would you prefer something like `pytools.{self.__class__}` or `{pytools.{self.__class__.__bases__[0]}}`  ?

File: pytools/__init__.py [hunk 3]
```diff
@@ -411,9 +411,16 @@ class RecordWithoutPickling:
     """
 
     __slots__: ClassVar[List[str]] = []
-    fields: ClassVar[Set[str]]
 
-    def __init__(self, valuedict=None, exclude=None, **kwargs):
+    # A dict, not a set, to maintain a deterministic iteration order
+    fields: ClassVar[Dict[str, None]]
```
Comment (matthiasdiener) [inline pytools/__init__.py:416]: Side note: this change will require mypy-related changes in loopy

File: pytools/__init__.py (hunk 1 above)
Comment (inducer) [reply pytools/__init__.py:471]: Yes, probably. But :shrug: not urgent, we're getting rid of this anyway.

**Comment by inducer** [comment]
I'll be honest, I underestimated the compatibility impact of this. I've just pushed a revert.

**Comment by inducer** [comment]
https://github.com/inducer/loopy/pull/785 gets rid of a good bunch more `Record`s in Loopy, which may make this more plausible.

**Comment by matthiasdiener** [comment]
Sure, no problem! Sorry for the confusion. Is there a particular error that came up with this?

**Comment by inducer** [comment]
It was code like this that made me substantially uneasy:

https://github.com/inducer/loopy/blob/d5512d5c6aec12f16a60aa5da31701c23d006bbd/loopy/kernel/instruction.py#L206-L211

**Pull request merged**